### PR TITLE
Functions for generating common quantum states

### LINKED ISSUE #146: Functions for generating quantum states
There are still some functions (compare to `qutip`) unsupported yet.

TODO:

- #150
  - [x] Move the existing functions and the following ones into an independent file `qobj/states.jl`
  - [x] set default value for argument `pos` in `basis` and `fock` as `0`

- #164
  - [x] `zero_ket`
  - [x] `fock_dm`
  - [x] `coherent_dm`
  - [x] `thermal_dm`
  - [x] `maximally_mixed_dm`
  - [x] `spin_state`
  - [x] `spin_coherent`
  - [x] `bell_state`
  - [x] `singlet_state`
  - [x] `triplet_states`
  - [x] `w_state`
  - [x] `ghz_state`

### PR DESCRIPTION
**Pull request by ytdHuang**
close #146 

Summarize of this PR:
- Introduce `zero_ket`
- Introduce `fock_dm`
- Introduce `coherent_dm`
- Introduce `thermal_dm`
- Introduce `maximally_mixed_dm`
- minor changes in `jmat`
- Introduce `spin_state`
- Introduce `spin_coherent`
- Introduce `bell_state`
- Introduce `singlet_state`
- Introduce `triplet_states`
- Introduce `w_state`
- Introduce `ghz_state`
- ignore size threshold for document API page: `size_threshold_ignore = ["api.md"]`

### CODE REVIEW

**Comment by ytdHuang** [comment]
All the code coverage update in runtests failed.

**Comment by ytdHuang** [comment]
okay, everything works fine again.

File: src/qobj/states.jl [hunk 1]
```diff
@@ -32,15 +46,68 @@ It is also possible to specify the list of dimensions `dims` if different subsys
...
+
+Density matrix for a thermal state (generating thermal state probabilities) with the following arguments:
+- `N::Int`: Number of basis states in the Hilbert space
+- `n::Real`: Expectation value for number of particles in the thermal state.
+"""
+function thermal_dm(N::Int, n::Real)
+    β = log(1.0 / n + 1.0)
+    N_list = Array{Float64}(0:N-1)
+    data = exp.(-β .* N_list)
+    return QuantumObject(spdiagm(0 => data ./ sum(data)), Operator, [N])
```
Comment (albertomercurio) [inline src/qobj/states.jl:93]: We should also have the case of both dense and sparse case.

File: src/qobj/states.jl [hunk 2]
```diff
@@ -54,3 +121,151 @@ function rand_dm(N::Integer; dims::Vector{Int} = [N])
...
+julia> bell_state("00")
+Quantum Object:   type=Ket   dims=[2, 2]   size=(4,)
+4-element Vector{ComplexF64}:
+ 0.7071067811865475 + 0.0im
+                0.0 + 0.0im
+                0.0 + 0.0im
+ 0.7071067811865475 + 0.0im
+```
+"""
+function bell_state(state::String = "00")
```
Comment (albertomercurio) [inline src/qobj/states.jl:198]: Why do we have a string as an argument? I think it is better to have two integer arguments.

File: src/qobj/states.jl (hunk 1 above)
Comment (ytdHuang) [reply src/qobj/states.jl]: okay

File: src/qobj/states.jl (hunk 2 above)
Comment (ytdHuang) [reply src/qobj/states.jl]: okay

**Review by albertomercurio — APPROVED** [review]

**Pull request merged**